// foster-backend/lib/user/userConstants.js

const MESSAGES = {
  passwordNotMatched: "Password is incorrect!",
  passwordChanged: "Password changed successfully",
  emailAlrdyRegistered: "An account already exists for this email.",
  invalidEmail: "Invalid Email",
  emailNotFound: "Email not found!",
  email_is_empty: "Please enter email.",
  registerUserMsg: "Account has been created and  We have emailed instructions to verify your email.",
  password_is_empty: "Please enter password.",
  nameEmpty: "Name cannot be empty.",
  lastNameEmpty: "Last name cannot be empty.",
  emailNoEmpty: "email cannot be empty.",
  genderEmpty: "Gender cannot be empty.",
  roleEmpty: "Role cannot be empty.",
  invalidRole: "Enter a valid role.",
  dobEmpty: "DOB cannot be empty.",
  otpSentTOMobile: "Verification link has been sent to your email.",
  emailNotExist: "email doesn't exist.",
  registeredUser: "User has been registered successfully.",
  invalidOtp: "Link is invalid.",
  otpExpired: "OTP is expired.",
  otpVerified: "OTP verified successfully.",
  otpSent: "Code sent successfully.",
  bucketCantEmpty: "Bucket name cannot be empty.",
  schoolCreated: "School created successfully.",
  bucketSelected: "Bucket selected successfully.",
  selectedBucketEmpty: "Selected bucket list cannot be empty.",
  atmost4Bucket: "Selected atmost four bucket.",
  schoolCantEmpty: "School cannot be empty.",
  ethnicityCantEmpty: "Ethnicity cannot be empty.",
  cityCantEmpty: "City cannot be empty.",
  stateCantEmpty: "State cannot be empty.",
  countryCantEmpty: "Country cannot be empty.",
  zipcodeCantEmpty: "Zipcode cannot be empty.",
  localityCantEmpty: "Locality area cannot be empty.",
  flatCantEmpty: "Flat no. cannot be empty.",
  fieldsCantEmpty: "Fields cannot be empty.",
  inActive: "Your account is temporarily blocked or deleted. Please contact site admin.",
  invalid_user: "Invalid email or password.",
  accessTokenCantEmpty: "Access token cannot be empty",
  settingForEmpty: "Please provide a setting key to set.",
  settingValueEmpty: "Please provide a setting value.",
  pagenameEmpty: "Please provide a page name.",
  pageContentEmpty: "Please provide page content.",
  pageAlreadyExists: "Page with the same name already exists.",
  pageSavedSuccess: "static page saved successfully.",
  referralCodeRedeemed: "Referral code redeemed successfully.",
  userForgotPasswordLinkSend: "We have sent a link via email to the respective user.",
  userIdEmpty: "User Id cannot be empty.",
  requestIdEmpty: "Request Id cannot be empty.",
  requestStatusEmpty: "Request status cannot be empty.",
  invalidRequestStatus: "Enter a valid Request status.",
  senderIdEmpty: "Sender Id cannot be empty.",
  childRemoveSuccess: "Child removed successfully.",
  parentRemoveSuccess: "Parent removed successfully.",
  pointsEmpty: "Please provide points to distribute",
  zeroPointNotAllowed: "You can not distribute 0 or less than 0 points",
  receiverIdEmpty: "Receiver Id cannot be empty.",
  insuffCoins: "Insufficient points",
  noUserFound: "No such User.",
  receiverNotSpotivity: "You can only distribute points to spotivity.",
  invalidCoinReceiver: "You can only distribute points to spotivity or your children.",
  noRelation: "No such Relation",
  zeroPagetNotAllowed: "Please enter the valid pageNo or PageSize",
  spotivtyCant: "You can not distribute",
  decimalPointNotAllowed: "Please enter a positive integer Points:",
  requestRejected: "Connection request rejected.",
  requestAccepted: "Connection request accepted successfully.",
  linkSent: "A verification link has been sent to your email. Please verify."


};

const ROLE = {
  USER: 1,
  BUSINESS: 2,
  ADMIN: 4
};

const OTP_EXPIRY_TIME = 1 * 60 * 1000; // 1 minute expire time

const USER_STATUS = {
  ACTIVE: 1,
  INACTIVE: 2,
  OTP_NOT_VERIFIED: 0,
  NEW_FB_USER: 3,
};

const SELECTED_MOST_BUCKET = 4;



const GAINED_REWARD_POINT_BY = {
  SIGN_UP: {
    BY: 1,
    POINT: 50,
    msg: "Spotivitiy's thank you for signing up! "
  },
  COMPLETE_PROFILE: {
    BY: 2,
    POINT: 15,
    msg: "You earned x points for successfully completing your spotivity profile"
  },
  REFERRAL_CODE: {
    BY: 3,
    POINT: 5, // default
    PARENT_GAURDIAN_POINT: 5,
    USER_CHILD: 10,
    msg: "You earned x points for a successful refferal "
  },
  CHECKIN: {
    BY: 4,
    POINT: 1,
    msg: "You earned x points thanks to checking-in"
  },
  SOCIAL_SHARE: {
    BY: 5,
    POINT: 1,
    msg: "You earned x points thanks to a social share"
  },
  SOCIAL_CONNECT: {
    BY: 6,
    POINT: 3,
    msg: "You earned x points for linking a social account to your profile"
  },
  DISTRIBUTION: {
    BY: 10,
    POINT: 3,
    msg: "Points gained by DISTRIBUTION"
  },
  INSIGHT_TEST: {
    BY: 11,
    POINT: 25,
    msg: "Points gained by Personal Insight Test"
  }
};

const COMPLETE_PROFILE_FIELDS_COUNT = 8;

const COMPLETE_PROFILE_FIELDS = ["school", "ethnicity",
  "state", "zipcode", "country", "locality", "flat", "city"];

const LOGIN_TYPE = {
  NORMAL_LOGIN: 1,
  FB: 2,
  INSTRAGRAM: 3,
  GOOGLE_PLUS: 4,
};

const APP_INVITE_TEMPLATE = {
  // MESSAGE: "[ReferralName] thinks you are awesome and wants to share Spotivity with you!  Spot Activities.  Earn Points.  Develop your passion.  Go explore and shape your future!",
  MESSAGE: "[ReferralName] thinks you are awesome and would like to invite you to download spotivity",
};

const REFERRAL_CODE = {
  SENT: 1,
  REDEEMED: 2,
};

const MEDIA_TYPE = {
  IMAGE: 1,
  ICON: 2,
};

const COUNTRY_STATE_CITY = {
  COUNTRY: 1,
  STATE: 2,
  CITY: 3,
  COUNTRY_ID: 231,
};

const MEMBER_STATUS = {
  USER_NOT_EXIST: 1,
  ALREADY_CHILD: 2,
  ALREADY_PARENT: 3,
  ONLY_A_USER: 4,
};

const REQUEST_MESSAGE = "You have received a request from [username] with contact number [phone] to be a part of the family"

const INSIGHT_TOOL_CATEGORY = {
  O_CATEGORY: "O",
  C_CATEGORY: "C",
  E_CATEGORY: "E",
  A_CATEGORY: "A",
  N_CATEGORY: "N",
}

const INSIGHT_TOOL_GRADE = {
  O_POSITIVE: "O+",
  O_POSITIVE_DESC: "You are high on openness. You possess flexibility in the way you think about the world and therefore enjoy a variety of feelings, experiences, and ideas. We recommend you engage in activities that are new to you or that might let you learn or do new things!",
  O_NEGATIVE: "O-",
  O_NEGATIVE_DESC: "You are low on openness. You are consistent in the way you think about the world and therefore appreciate practical and familiar feelings, experiences, and ideas. We recommend you engage in activities that you are good at and enjoy!",
  C_POSITIVE: "C+",
  C_POSITIVE_DESC: "You are high on conscientiousness. You are organized, persistent, and disciplined in pursuing your goals. We recommend you engage in activities that allow you to utilize your problem-solving skills and knowledge!",
  C_NEGATIVE: "C-",
  C_NEGATIVE_DESC: "You are low on conscientiousness. You are sometimes unorganized in, or unmotivated to follow through on, things you set out to do. We recommend you engage in “outside-the-box” activities that allow freedom and autonomy to learn!",
  E_POSITIVE: "E+",
  E_POSITIVE_DESC: "You are high on extraversion. You have a need to engage in stimulating activities that afford frequent and intense social interactions. We recommend you engage in activities that allow you to meet new people and build meaningful relationships!",
  E_NEGATIVE: "E-",
  E_NEGATIVE_DESC: "You are low on extraversion. You enjoy engaging in individual activities that afford autonomy and personal space. We recommend you engage in activities that allow you to work independently and hone a new or existing passion!",
  A_POSITIVE: "A+",
  A_POSITIVE_DESC: "You are high on agreeableness. You are compassionate and trusting rather than antagonistic when interacting with others. We recommend you engage in activities that allow you to build collaborative friendships in the service of others!",
  A_NEGATIVE: "A-",
  A_NEGATIVE_DESC: "You are low on agreeableness. You are suspicious in new situations, and at times antagonistic when interacting with others in a social setting. We recommend you engage in activities that allow you to use your intellect to find answers to difficult problems.",
  N_POSITIVE: "N+",
  N_POSITIVE_DESC: "You are high on neuroticism. Your emotions vary widely across differing social situations, which can lead to anxiety. We recommend you engage in activities that challenge you but allow you to feel comfortable with your and others’ goals and actions!",
  N_NEGATIVE: "N-",
  N_NEGATIVE_DESC: "You are low on neuroticism. Your emotions are fairly stable, which leads you to be calm across a range of social situations. We recommend you engage in personally satisfying and collaborative activities with other individuals!",
}
//= ========================= Export Module Start ===========================

module.exports = {
  MESSAGES,
  ROLE,
  OTP_EXPIRY_TIME,
  USER_STATUS,
  SELECTED_MOST_BUCKET,
  GAINED_REWARD_POINT_BY,
  COMPLETE_PROFILE_FIELDS_COUNT,
  COMPLETE_PROFILE_FIELDS,
  LOGIN_TYPE,
  APP_INVITE_TEMPLATE,
  REFERRAL_CODE,
  MEDIA_TYPE,
  COUNTRY_STATE_CITY,
  MEMBER_STATUS,
  REQUEST_MESSAGE,
  INSIGHT_TOOL_CATEGORY,
  INSIGHT_TOOL_GRADE
};

//= ========================= Export Module END ===========================
