// foster-backend/lib/model/users.js
// Importing mongoose
const mongoose = require("mongoose");

const Schema = mongoose.Schema;
const constants = require("../constants");

let User;

const UserSchema = new Schema(
	{
		email: {
			type: String,
			required: false,
			index: true,
			unique: true,
			sparse: true,
			lowercase: true,
			trim: true,
		},
		password: { type: String },
		role: {
			type: Number, // 1 - parent, 2 - student
		},
		name: {
			type: String,
			trim: true,
		},
		created_at: {
			type: Number,
			default: Date.now(),
		},
		updated_at: {
			type: Number,
			default: Date.now(),
		},
		noti_setting: {
			type: Boolean,
			default: true,
		},
		child_ids: [{ type: Schema.Types.ObjectId }],
		registration_ids: [
			{
				id: { type: String },
				plat_form: { type: Number },
				snsEndpoint: { type: String }
			}
		],
		reset_password_token: {
			type: String,
			trim: true,
		},
		status: {
			type: Number,
			default: 0
		},
		otp: {
			type: Number,
		},
		profile_pic: {
			type: String,
			default: ""
		},
		address: {
			type: String,
			default: ""
		},
		latitude: {
			type: Number
		},
		longitude: {
			type: Number
		}
	},
	{
		versionKey: false,
	},
);

UserSchema.methods.toJSON = function () {
	const obj = this.toObject();
	delete obj.password;
	return obj;
};

// Export user module
User = module.exports = mongoose.model("User", UserSchema);
