// foster-backend/lib/model/userFolders.js
// Importing mongoose
const mongoose = require("mongoose");

const Schema = mongoose.Schema;
const constants = require("../constants");

let UserFolder;

const UserFolderSchema = new Schema(
    {
        folder_name: {
            type: String,
            required: true,
            trim: true,
        },
        created_at: {
            type: Number,
            default: Date.now(),
        },
        updated_at: {
            type: Number,
            default: Date.now(),
        },
        parentFolderId: { type: Schema.Types.ObjectId },
        userId: { type: Schema.Types.ObjectId },
        images: [
            {
                url: { type: String },
            }
        ],
        is_default: {
            type: Boolean,
            default: false
        }
    },
    {
        versionKey: false,
    },
);

UserFolderSchema.methods.toJSON = function () {
    const obj = this.toObject();
    delete obj.password;
    return obj;
};

// Export user module
UserFolder = module.exports = mongoose.model("UserFolder", UserFolderSchema);
