// foster-backend/lib/dao/baseDao.js


//= ========================= Class Definitions Start =====================

class BaseDao {
  constructor(dbModel) {
    // Get Model
    this.Model = dbModel;
  }

  save(object) {
    return this.Model.create(object);
  }

  findOne(query, projection) {
    return this.Model.findOne(query, projection).exec();
  }

  find(query, projection) {
    return this.Model.find(query, projection).exec();
    /* var query = this.Model.find(query, projection);
     query.limit(10);
     if (options.limit) {
     query.limit(options.limit);
     }
     return query.exec(); */
  }

  findOneAndUpdate(query, update, options) {
    return this.Model.findOneAndUpdate(query, update, options).exec();
  }

  aggregate(aggPipe) {
    console.log("aggPipe",aggPipe)
    return this.Model.aggregate(aggPipe).exec();
  }



}
//= ========================= Class Definitions End =====================


//= ========================= Helper methods start =======================

//= ========================= Helper methods end =======================

//= ========================= Export module start =======================

module.exports = BaseDao;

//= ========================= Export module End =======================
