// foster-backend/lib/constants.js
const MESSAGES = {
  keywordNotMatched: "Sorry we were not able to find that word anywhere",
  requiredField: "Required field are missing.",
  intrnlSrvrErr: "Please try after some time.",
  emailAlrdyRegistered: "An account already exists for this email/phone.",
  invalidEmail: "Invalid Email",
  registerUserMsg: "Account has been created and  We have emailed instructions to verify your email.",
  tokenGenError: "Unable to generate token",
  invalidMobileNo: "Invalid phone number",
  emailOrPhoneDoesNotExists: "Email or phone does not exist.",
  linkExpired: "Link has been expired.",

};

const DB_MODEL_REF = {
  USER: "User",
  SOCIAL_NETWORKS: "social_networks",
  CATEGORIES: "category",
  TAGS: "tag",
  PRODUCTS: "product"

};

const STATUS_CODE = {
  ERROR: 0,
  SUCCESS: 1,
};

const GENDER = {
  MALE: { KEY: "male", VALUE: 1 },
  FEMALE: { KEY: "female", VALUE: 2 },
  NEITHER: { KEY: "neither", VALUE: 3 },
  BOTH: { KEY: "both", VALUE: 4 },
};

const ROUTE_PREFIX = {
  NEW_BASE_PATH: "/foster/",
};

const otpMailBody = `Foster would like to verify your account for security purposes.\n[OTP] is your OTP code to securely activate your Foster account.\nEnjoy searching for activities and finding those that help you develop your passion.`


//= ========================= Export Module Start ===========================

module.exports = {
  MESSAGES,
  ROUTE_PREFIX,
  DB_MODEL_REF,
  STATUS_CODE,
  GENDER,
  otpMailBody
};

//= ========================= Export Module END ===========================
