// foster-backend/lib/config/env/default.js

console.log("default.js");

//= ========================= Class Definitions Start =====================

//= ========================= Export module end =========================

class DefaultConfig {
  constructor() {
    this.TAG = "local";
    this.isProd = false;
    this.port = 3006;
    // Server IP
    this.hostIP = "127.0.0.1";
    this.hostName = "localhost";
    this.JWT_TOKEN_SECRET = "bf23d39ca08631ea94370d597324338b";
    this.webBaseUrl = "http://localhost:3002/";

    this.mongo = {
      dbName: "foster",
      dbUrl: "mongodb://localhost:27017/",
      options: {
        user: "foster",
        pass: "foster",
        server: { socketOptions: { keepAlive: 1, connectTimeoutMS: 30000 } },
        replset: { socketOptions: { keepAlive: 1, connectTimeoutMS: 30000 } },
      },
    };
    this.redisDb = {
      redisDBIndex: 0,
      port: 6379,
      host: "127.0.0.1",
    };
    this.s3_credentials = {
      accessKey: "fdBjoDkYWqIuUzWy+RF9PUgvJjLZ/Xe8qmrczSYc",
      keyId: "AKIAIN4CUTMGD7VDU52Q",
    };

    // option parameters constants for s3
    this.s3 = {
      maxAsyncS3: 20, // this is the default
      s3RetryCount: 3, // this is the default
      s3RetryDelay: 1000, // this is the default
      multipartUploadThreshold: 20971520, // this is the default (20 MB)
      multipartUploadSize: 15728640, // this is the default (15 MB)
      bucketName: "spotivity",
      publicBucketName: "spotivity",
      signatureVersion: "v2",
      region: "us-east-1",
      ACL: "public-read",
    };

    this.sendGrid = {
      apiKey: "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",
      FROM: "admin@foster.com",
    };

  }



  getConfigs() {
    return this;
  }
}

//= ========================= Class Definitions End =====================


//= ========================= Export module start =======================

module.exports = DefaultConfig;

//= ========================= Export module end =========================
