// foster-backend/lib/config/env/dev.js


console.log("dev.js");
//= ========================= Load Modules Start ==========================

//= ========================= Load Internal Modules =======================

const DefaultConfig = require("./default");

//= ========================= Load Modules End ============================

//= ========================= Class Definitions Start =====================

class DevConfig extends DefaultConfig {
  constructor() {
    super();

    // override dev property
    this.TAG = "development";
    this.port = 3500;
    this.redisDb = {
      redisDBIndex: 0,
      port: 6379,
      host: "127.0.0.1",
    };
    this.mongo = {
      dbName: "foster",
      dbUrl: "mongodb://localhost:27017/",
      options: {
        user: "foster",
        pass: "foster@123",
        server: { socketOptions: { keepAlive: 1, connectTimeoutMS: 30000 } },
        replset: { socketOptions: { keepAlive: 1, connectTimeoutMS: 30000 } },
      },
    };
   
  }
}

//= ========================= Class Definitions End =======================

module.exports = DevConfig;
