// foster-backend/lib/config/env/local.js


console.log("local.js");

//= ========================= Load Modules Start ==========================

//= ========================= Load Internal Modules =======================

const DefaultConfig = require("./default");

//= ========================= Load Modules End ============================

//= ========================= Class Definitions Start =====================

class LocalConfig extends DefaultConfig {
  constructor() {
    super();
    this.port = 3008;
    this.TAG = "local";
    this.baseURL = "localhost:3001";
    this.webBaseUrl = "http://localhost:3002/";
    this.mongo = {
      dbName: "foster",
      dbUrl: "mongodb://localhost:27017/",
      options: {
        // user: "foster",
        // pass: "foster",
        server: { socketOptions: { keepAlive: 1, connectTimeoutMS: 30000 } },
        replset: { socketOptions: { keepAlive: 1, connectTimeoutMS: 30000 } },
      },
    };
    this.redisDb = {
      redisDBIndex: 0,
      port: 6379,
      host: "127.0.0.1",
    };
   
  }
}

//= ========================= Class Definitions End =======================

module.exports = LocalConfig;
