// foster-backend/lib/config/env/prod.js


//= ========================= Load Modules Start ==========================

//= ========================= Load Internal Modules =======================

const DefaultConfig = require("./default");

//= ========================= Load Modules End ============================

//= ========================= Class Definitions Start =====================

class ProdConfig extends DefaultConfig {
  constructor() {
    super();
    this.port = 3006;
    this.hostIP = "18.204.161.36";
    this.TAG = "production";
    this.COUNTRY_CODE = "+1";
    this.redisDb = {
      redisDBIndex: 0,
      port: 6379,
      host: "127.0.0.1",
    };
    this.mongo = {
      dbName: "foster",
      dbUrl: "mongodb://localhost:27018/",
      options: {
        user: "foster",
        pass: "foster",
        server: { socketOptions: { keepAlive: 1, connectTimeoutMS: 30000 } },
        replset: { socketOptions: { keepAlive: 1, connectTimeoutMS: 30000 } },
      },
    };
    
  }
}

//= ========================= Class Definitions End =======================


module.exports = ProdConfig;
