// foster-backend/lib/config/env/staging.js


//= ========================= Load Modules Start ==========================

//= ========================= Load Internal Modules =======================

const DefaultConfig = require("./default");

//= ========================= Load Modules End ============================

//= ========================= Class Definitions Start =====================

class StagingConfig extends DefaultConfig {
  constructor() {
    super();
    this.TAG = "staging";
    this.COUNTRY_CODE = "+1";
    this.isProd = false;
    this.port = 3006;
    // Server IP
    this.hostIP = "18.211.82.47";
    this.hostName = "staging";
    this.webBaseUrl = "https://stag.spotivity.com";
    this.redisDb = {
      redisDBIndex: 0,
      port: 6379,
      host: "127.0.0.1",
    };
    this.mongo = {
      dbName: "foster",
      dbUrl: "mongodb://localhost:27018/",
      options: {
        user: "foster",
        pass: "foster",
        server: { socketOptions: { keepAlive: 1, connectTimeoutMS: 30000 } },
        replset: { socketOptions: { keepAlive: 1, connectTimeoutMS: 30000 } },
      },
    };
    this.RESET_PASSWORD = {
      LINK: "http://dev.algoworks.com/foster-web-dev/resetPassword/",
    };


  }
}

//= ========================= Class Definitions End =======================

module.exports = StagingConfig;
